# Paul the Octopus

In this notebook, we will analyze data from previous FIFA World Cup matches and try using it to predict the outcome of future matches.

## Dataset

First, we import pandas library.

In [2]:
import pandas as pd

We then read the datasets we will analyze. These datasets were obtained from [kaggle](https://www.kaggle.com/abecklas/fifa-world-cup/) and contain information about the matches, players and cups.

In [3]:
matches = pd.read_csv('data/WorldCupMatches.csv')
players = pd.read_csv('data/WorldCupPlayers.csv')
world_cups = pd.read_csv('data/WorldCups.csv')

Let's have a look at the characteristics of each dataset imported.

### Matches dataset

As we can see below, this dataset contains general information about every World Cup match. It depicts the date, location and attendance of the game, the stage (group or knockout round), name and initials of home and away teams, half-time and final score, win conditions (decided in extra time or penalties), name of referees and round and  match ids.

In [6]:
matches.head(3)

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                     4444.0                   3.0                   0.0   
1                    18346.0                   2.0                   0.0   
2                    24059.0                   2.0                   0.0   

                  Referee               Assistant 1            Assistant 2  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)    REGO Gilberto (BRA)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)  WARNKEN Alberto (CHI)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)    BALWAY Thomas (FRA)   

   RoundID  MatchID Home Team Initials Away Team Initials  
0    201.0   1096.0                FRA                MEX  
1    201.0   1090.0                USA                BEL  
2    201.0   1093.0                YUG                BRA

Now, let's check the size of our dataset and if there is any empty values in it.

In [20]:
matches.shape

(4572, 20)

In [21]:
matches.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4572 entries, 0 to 4571
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Year                  852 non-null    float64
 1   Datetime              852 non-null    object 
 2   Stage                 852 non-null    object 
 3   Stadium               852 non-null    object 
 4   City                  852 non-null    object 
 5   Home Team Name        852 non-null    object 
 6   Home Team Goals       852 non-null    float64
 7   Away Team Goals       852 non-null    float64
 8   Away Team Name        852 non-null    object 
 9   Win conditions        852 non-null    object 
 10  Attendance            850 non-null    float64
 11  Half-time Home Goals  852 non-null    float64
 12  Half-time Away Goals  852 non-null    float64
 13  Referee               852 non-null    object 
 14  Assistant 1           852 non-null    object 
 15  Assistant 2          

So, it looks like we have 4572 rows, but only 852 of them are non-null. Also, there's 2 more values missing in the **Attendance** column. We'll treat those missing values later.

### Players dataset

This dataset contains data of the players that were in the line-up of all matches. We can see the player's name, team, position (goal keeper or captain), coach name and shirt number. Also, there is information about wheter the player was in line-up or substitute and any event (goal, card, substitution) related to the player that occurred in the match. Finally, matches and rounds are identified by their unique ids.

In [6]:
players.head(3)

RoundID  MatchID Team Initials           Coach Name Line-up  Shirt Number  \
0      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   
1      201     1096           MEX     LUQUE Juan (MEX)       S             0   
2      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   

        Player Name Position Event  
0       Alex THEPOT       GK   NaN  
1   Oscar BONFIGLIO       GK   NaN  
2  Marcel LANGILLER      NaN  G40'

It seems our dataset only has empty cells in the last two columns (**Position** and **Event**). This is ok, since only goalkeepers and captains are identified in the former and not all players were expected to be involved in events during a match.

In [25]:
players.shape

(37784, 9)

In [23]:
players.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37784 entries, 0 to 37783
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   RoundID        37784 non-null  int64 
 1   MatchID        37784 non-null  int64 
 2   Team Initials  37784 non-null  object
 3   Coach Name     37784 non-null  object
 4   Line-up        37784 non-null  object
 5   Shirt Number   37784 non-null  int64 
 6   Player Name    37784 non-null  object
 7   Position       4143 non-null   object
 8   Event          9069 non-null   object
dtypes: int64(3), object(6)
memory usage: 2.6+ MB


### World Cups dataset

The final dataset contains general data about the world cup editions, such as the year it occurred, hosting country/countries, winner, 2nd, 3rd and 4th places, number of goals scored, participants, matches played and overall attenndance.

In [8]:
world_cups.head(3)

Year  Country   Winner      Runners-Up    Third      Fourth  GoalsScored  \
0  1930  Uruguay  Uruguay       Argentina      USA  Yugoslavia           70   
1  1934    Italy    Italy  Czechoslovakia  Germany     Austria           70   
2  1938   France    Italy         Hungary   Brazil      Sweden           84   

   QualifiedTeams  MatchesPlayed Attendance  
0              13             18    590.549  
1              16             17    363.000  
2              15             18    375.700

In [27]:
world_cups.shape

(20, 10)

In [28]:
world_cups.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Year            20 non-null     int64 
 1   Country         20 non-null     object
 2   Winner          20 non-null     object
 3   Runners-Up      20 non-null     object
 4   Third           20 non-null     object
 5   Fourth          20 non-null     object
 6   GoalsScored     20 non-null     int64 
 7   QualifiedTeams  20 non-null     int64 
 8   MatchesPlayed   20 non-null     int64 
 9   Attendance      20 non-null     object
dtypes: int64(4), object(6)
memory usage: 1.7+ KB


So, let's say we would like to know how many World Cup matches the brazilian national team has played since its first edition, in 1930.

We must first filter the original dataframe **df1** by teams, and then get the size of the filtered dataframe.

In [10]:
df_br = df1[df1['Home Team Name'] == 'Brazil']

print('The brazilian team has played %d matches since 1930 FIFA World Cup.' % len(df_br))

The brazilian team has played 82 matches since 1930 FIFA World Cup


82 matches seems a lot, but actually we are just considering the matches where Brazil was set as the home team.

To obtain the real number of matches, we must also consider the matches where Brazil was set as away team. Let's do it.

In [13]:
df_br = df1[(df1['Home Team Name'] == 'Brazil') | (df1['Away Team Name'] == 'Brazil')]

print('The real number of matches played by the brazilian team is %d.' % len(df_br))

The real number of matches played by the brazilian team is 108


The operand `|` stands for a logic **or**, meaning that we want all rows that have Brazil on the fields `Home Team Name` or `Away Team Name`.

Obs.: Notice that each filter was inside `()`. This is needed when applying multiple conditions.

## Highest and Lowest Attendances

We are now interested to know which were the games with most and least public attendance. To do so, we must get the maximum and minimum value present on the column `Attendance` of our dataframe.

In [23]:
max_att = df1['Attendance'].max()
min_att = df1['Attendance'].min()

print('The highest attendance on a WC match was %d people, whilst the '
      'lowest attendance was only %d people' % (max_att, min_att))

The highest attendance on a WC match was 173850 people, whilst the lowest attendance was only 2000 people


However, we would also like to know in what games those attendances occurred. First, let's find in what rows of our original dataframe these attendances are located.

In [36]:
id_max_att = df1['Attendance'].idxmax()
id_min_att = df1['Attendance'].idxmin()

print('The highest and lowest attendances appear on rows #%d and #%d, respectively' % (id_max_att, id_min_att))

The highest and lowest attendances appear on rows #74 and #9, respectively


Knowing the row's id, we can access the data contained in it using functions `loc` or `iloc`. Basically, the difference between those function is how the user address the columns, with the first requiring the full name of the column while the other receives the id of the columns of interest.

In [48]:
date_max = df1.iloc[id_max_att, 1]
venue_max = df1.iloc[id_max_att, 3]
city_max = df1.iloc[id_max_att, 4]
home_team_max = df1.iloc[id_max_att, 5]
away_team_max = df1.iloc[id_max_att, 8]
score_home_max = df1.iloc[id_max_att, 6]
score_away_max = df1.iloc[id_max_att, 7]
short_home_max = df1.iloc[id_max_att, 18]
short_away_max = df1.iloc[id_max_att, 19]

print('The highest attendance in a FIFA World Cup match was registered in %s at %s in %s.' 
      % (date_max, venue_max, city_max))
print('The match was between %s and %s and the final score was (%s) %d x %d (%s)' 
      % (home_team_max, away_team_max, short_home_max, score_home_max, score_away_max, short_away_max))

The highest attendance in a FIFA World Cup match was registered in 16 Jul 1950 - 15:00  at Maracan� - Est�dio Jornalista M�rio Filho in Rio De Janeiro .
The match was between Uruguay and Brazil and the final score was (URU) 2 x 1 (BRA)


In [52]:
date_min = df1.loc[id_min_att, 'Datetime']
venue_min = df1.loc[id_min_att, 'Stadium']
city_min = df1.loc[id_min_att, 'City']
home_team_min = df1.loc[id_min_att, 'Home Team Name']
away_team_min = df1.loc[id_min_att, 'Away Team Name']
score_home_min = df1.loc[id_min_att, 'Home Team Goals']
score_away_min = df1.loc[id_min_att, 'Away Team Goals']
short_home_min = df1.loc[id_min_att, 'Home Team Initials']
short_away_min = df1.loc[id_min_att, 'Away Team Initials']

print('The lowest attendance in a FIFA World Cup match was registered in %s at %s in %s.' 
      % (date_min, venue_min, city_min))
print('The match was between %s and %s and the final score was (%s) %d x %d (%s)' 
      % (home_team_min, away_team_min, short_home_min, score_home_min, score_away_min, short_away_min))

The lowest attendance in a FIFA World Cup match was registered in 19 Jul 1930 - 12:50  at Estadio Centenario in Montevideo .
The match was between Chile and France and the final score was (CHI) 1 x 0 (FRA)


<kbd>Alt<\